src/js/components/conversation.jsx

```diff
@@ -54,34 +65,145 @@ export class ConversationComponent extends Component {
+
+        const node = findDOMNode(this);
+        if(hasMoreMessages && !isFetchingMoreMessages) {
+            // make sure the last message is one from the server, otherwise it doesn't need to scroll to previous first message
+            if (messages.length > 0 && messages[messages.length - 1]._id) {
+                this._lastTopMessageId = messages[0]._id;
+            }
+            
+            const top = getTop(this._topMessageNode, node);
+            this._lastTopMessageNodePosition = top - node.scrollTop;
+            dispatch(setFetchingMoreMessages(true));
+            setTimeout(() => {
+                fetchMoreMessages();
+            }, 400);
+        }
+    };
+
     scrollToBottom = () => {
-        const timeout = setTimeout(() => {
+        const {shouldScrollToBottom} = this.props;
+        if(!this._isScrolling && (shouldScrollToBottom || this._forceScrollToBottom)) {
+            this._isScrolling = true;
+            const timeout = setTimeout(() => {
+                const container = findDOMNode(this);
+                const logo = this.refs.logo;
+                const scrollTop = container.scrollHeight - container.clientHeight - logo.clientHeight - INTRO_BOTTOM_SPACER;
+                container.scrollTop = scrollTop;
+                this._isScrolling = false;
+                this._forceScrollToBottom = false;
+            });
+            this.scrollTimeouts.push(timeout);
+        }
+    };
+
+    scrollToPreviousFirstMessage = () => {
+        if(this._lastTopMessageNodePosition && !this._isScrolling) {
             const container = findDOMNode(this);
-            const logo = this.refs.logo;
-            const scrollTop = container.scrollHeight - container.clientHeight - logo.clientHeight - INTRO_BOTTOM_SPACER;
-            container.scrollTop = scrollTop;
-        });
-        this.scrollTimeouts.push(timeout);
+            const node = this._lastTopMessageNode;
+            this._isScrolling = true;
+
+            // When fetching more messages, we want to make sure that after 
+            // render, the messages stay in the same places
+            container.scrollTop = getTop(node, container) - this._lastTopMessageNodePosition;
+
+            const timeout = setTimeout(() => {
+                this._isScrolling = false;
+            });
+
+            this.scrollTimeouts.push(timeout);
+            this._lastTopMessageNode = undefined;
+        }
     };
 
+    componentWillUpdate(nextProps) {
+        const {introHeight: currentIntroHeight, messages: currentMessages, isFetchingMoreMessages} = this.props;
+        const {introHeight: newIntroHeight, messages: newMessages} = nextProps;
+
+        if (currentIntroHeight !== newIntroHeight) {
+            this._forceScrollToBottom = true;
+        }
+
+        // Check for new appMaker messages
+        const isAppMakerMessage = newMessages.length - currentMessages.length === 1 ? newMessages.slice(-1)[0].role === 'appMaker' : false;
```

Comment:
`.role !== 'appUser'` is a better check. A message with role `whisper` is still technically an app maker message